esp/services/ws_workunits/ws_workunitsHelpers.cpp

```diff
@@ -2092,52 +2055,75 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
     OwnedIFileIOStream ios = createBufferedIOStream(rIO);
     if (slaveNum > 0)
     {
-        StringBuffer line;
-        bool eof = false;
-        bool include = false;
+        readWorkunitLog(ios.get(), buf, outFile);
+    }
+    else
+    {//legacy wuid
+        appendIOStreamContent(buf, ios.get(), forDownload);
+    }
+}
 
-        VStringBuffer startwuid("Started wuid=%s", wuid.str());
-        VStringBuffer endwuid("Finished wuid=%s", wuid.str());
+void WsWuInfo::readWorkunitLog(OwnedIFileIOStream ios, MemoryBuffer& buf, const char* outFile)
+{
+    VStringBuffer startwuid("Started wuid=%s", wuid.str());
+    VStringBuffer endwuid("Finished wuid=%s", wuid.str());
 
-        const char *sw = startwuid.str();
-        const char *ew = endwuid.str();
+    const char *sw = startwuid.str();
+    const char *ew = endwuid.str();
 
-        Owned<IFileIOStream> outIOS;
-        if (!isEmptyString(outFile))
-        {
-            CWsWuFileHelper helper(nullptr);
-            outIOS.setown(helper.createIOStreamWithFileName(outFile, IFOcreate));
-        }
-        while (!eof)
+    bool eof = false;
+    bool outputThisLine = false;
+
+    CWsWuFileHelper helper(nullptr);
+    Owned<IFileIOStream> outIOS = helper.createIOStreamWithFileName(outFile, IFOcreate);
+    while (!eof)
+    {
+        StringBuffer line;
+        for (;;)
         {
-            line.clear();
-            for (;;)
+            char c;
+            size32_t numRead = ios->read(1, &c);
```

Comment:
this is not going to be efficient.
If there isn't already, should add a utility function to jlib somewhere that reads a line (efficiently) from a stream.
e..g reads a buffer at a time.

Reply:
ok